firedrake/linear_solver.py

```diff
@@ -56,6 +56,7 @@ def __init__(self, A, *, P=None, solver_parameters=None,
                                                        ksp_defaults=self.DEFAULT_KSP_PARAMETERS)
         self.A = A
         self.comm = A.comm
+        self._comm = internal_comm(self.comm)
```

Comment:
Remind me why you carry both of these around? Why not just have `self.comm = internal_comm(A.comm)` everywhere and `decref` that? We've never promised that the comm you get out of firedrake objects is a "user" one.

Reply:
I agree, we never made such a promise. However, we do have documentation suggesting that users can use the `.comm` attribute, and it _is_ a public attribute. The main aim of this and the previous PR is to be more consistent with the comms that we hold references to since...
- PETSc garbage is now stored as a comm attribute
- previously it was all too easy to exhaust the number of available communicators.